### ISSUE TITLE
LSTM recurrent kernel ( hidden state ) makes Host to GPU transaction for every sequence, which looks quite inefficient.

### ISSUE THREAD

**Issue by boiihong** [tensorflow/tensorflow#26500]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes.
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 16.04.5 LTS
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): build from source
- TensorFlow version (use command below): v1.13.0-7-g9ca8321 1.12.0
- Python version: Python 3.6.1 :: Anaconda 4.4.0 (64-bit)
- Bazel version (if compiling from source): Build label: 0.19.2
- GCC/Compiler version (if compiling from source): gcc version 5.4.0 (GCC)
- CUDA/cuDNN version: 
I got cuda version by checking nvcc verison. it is Cuda compilation tools, release 10.0, V10.0.130.
cudnn 7.5.0
- GPU model and memory: v100

I am evaluating the inference performance of 2-level stacked LSTM model. 
As you could see in the code below, my model has 2 LSTM cell and one time-distributed (FC) layer.
Because i want to make close investigation of how inference works, i utilize google timeline tool.

**Describe the current behavior**
Whenever new sequence starts, the recurrent kernel ( hidden state of lstm cell ) makes Host to GPU data transaction. Because this data transfer is initiated by ReadVariableOp which is connected to matmul op of a lstm hidden state, i am pretty sure that a data is the weight of the hidden state.
What i am curious is that why the weight has to move from H2D? if right, where is the D2H transaction? Doesn't hidden state be renewed every new sequence?

**Describe the expected behavior**
I think there should be two kinds of correct behavior.
1. For every sequence of lstm, GPU renew the weight, and this is moved to Host memory (D2H transaction - which doesn't exist at the timeline result). And then the weight goes to GPU again at the next sequence. 
2. No Transaction. the hidden state is renewed and reused only inside the GPU memory. In this case, i could see the H2D transaction (which is observed at the timeline result)

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

```python
import tensorflow as tf
import tensorflow.keras.backend as K
import time

from tensorflow.keras.models import Sequential, Model
from tensorflow.keras.layers import Dense, Activation, LSTM, RepeatVector, Input, TimeDistributed
from tensorflow.python.client import timeline

# configuration
iteration = 1
batch = 1024
seq_len = 26
fvec = 256#1024
unit = 256 
dtype = tf.float32
np_dtype = 'float32'

# set precision of model
K.set_floatx(np_dtype)
print(K.floatx())

# session setting
K.clear_session()
config = tf.ConfigProto() 
sess = tf.Session(config=config)
K.set_session(sess)

# make random input data
inp_dims = ( batch, seq_len, fvec)
input_data = np.random.random_sample(inp_dims).astype(np_dtype)

with tf.device('/device:GPU:0'):
	input_box = tf.placeholder(dtype, shape=(None, seq_len,fvec))
	out = LSTM(unit  , activation='sigmoid', input_shape=(seq_len, fvec),  implementation=2)(input_box) # 1st LSTM
	out = RepeatVector(seq_len)(out)
	out = LSTM(unit, activation='sigmoid', return_sequences=True, implementation=2)(out)
	out = TimeDistributed(Dense(fvec))(out)

# variable initialization
init = tf.global_variables_initializer()
sess.run(init)

# for tracing
run_options = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)
run_metadata = tf.RunMetadata()

# check time lapse
start_time = time.time()
for i in range(iteration):
  output = sess.run(out, feed_dict={input_box:input_data} , options = run_options , run_metadata= run_metadata)
  print (type(output), output.shape)
end_time = time.time()

# get trace information
tl= timeline.Timeline(run_metadata.step_stats)
ctf = tl.generate_chrome_trace_format(show_memory=True)

# store the trace information
with open ("./all_model_trace_" + np_dtype +"_infer_0306_cpumem.json", 'w') as f:
  f.write(ctf)

#save graph file..
sess = K.get_session()
graph_writer = tf.summary.FileWriter("./output_4/graph", sess.graph)
tf.train.write_graph(sess.graph, './output_4/', 'saved_model.pb', as_text=True)

# print time lapse
time_lapse = end_time - start_time
print(time_lapse)
```

**Comment by alvaroslm** [comment]
You should try [CuDNNLSTM](https://www.tensorflow.org/api_docs/python/tf/keras/layers/CuDNNLSTM)

**Comment by boiihong** [comment]
Fast Evaluation reveals that CuDNNLSTM makes performance three times better than a simple LSTM layer. I'll compare the performance of FP32(Cudacore) and FP16(Tensorcore) and  post it later.

**Comment by ymodak** [comment]
Closing this issue due to lack of updates. Feel free to reopen if have any follow up questions. Thanks!
